File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;

class Solution {
    public String solution(String number, int k) {
        
        Deque<Integer> que = new ArrayDeque<>();
        int size = k;
        
        for(int i=0; i<number.length(); i++){
            int now = number.charAt(i) - '0';
            while(!que.isEmpty() && size > 0){
                if(que.peekLast() < now){
                    que.pollLast();
                    size--;
                }else break;
            }
            que.add(now);
        }
        
        StringBuilder sb = new StringBuilder();
        
        for(int i=0; i<number.length()-k ; i++){
            sb.append(que.poll());
        }
        
        return sb.toString();
    }
}

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.js
function solution(number, k) {
    
    const stack = []
    var size = k
    
    for(let i=0; i<number.length; i++){
        while(stack.length > 0 && stack[stack.length-1] < number.charAt(i)-'0' && k>0){
            stack.pop()
            k--
        }
        stack.push(number.charAt(i) - '0')
    }
    
    var answer = "";
    for(let i=0; i<number.length - size; i++){
        answer = answer + stack[i]
    }
    
    return answer;
}